File: Coding_Problem/Coding_Problem_In_Csharp/BestTimeToBuyAndSellStock.cs
using System.Collections.Generic;

namespace Coding_Problem_In_Csharp {
    public class BestTimeToBuyAndSellStock {
        private int minPrice = 10001;
        private int maxProfit = 0;

        public int Calculate(IEnumerable<int> prices) {
            foreach (var price in prices) {
                minPrice = min(minPrice, price);
                maxProfit = max(maxProfit, price - minPrice);
            }

            return maxProfit;
        }

        int min(int n1, int n2) => (n1, n2) switch {
            var a when a.n1 < a.n2 => n1,
            var a when a.n1 > a.n2 => n2,
            _ => n1
        };

        int max(int n1, int n2) => (n1, n2) switch {
            var a when a.n1 > a.n2 => n1,
            var a when a.n1 < a.n2 => n2,
            _ => n1
        };
    }
}

File: Coding_Problem/Python_Prac_Coding_Problem/BestTimeToBuyAndSellStock.py
# LeetCode - 121. Best Time to Buy and Sell Stock
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

def buyAndSell(self, prices: List[int]) -> int:
    min_price, max_profit = prices[0], 0
    for price in prices:
        min_price = min(min_price, price)
        max_profit = max(max_profit, price - min_price)
    return max_profit


buyAndSell([7, 1, 5, 3, 6, 4])
